File: bindings/wasm/audit_trail_wasm/examples/src/07_access_read_only_methods.ts
// Copyright 2026 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

/**
 * ## Actors
 *
 * - **Admin client**: Creates the trail and sets up the RecordAdmin role.
 * - **Record admin client**: Adds one follow-up record. All subsequent operations are read-only
 *   and can be performed by any address — no capability required.
 *
 * Demonstrates how to:
 * 1. Load the full on-chain trail object.
 * 2. Inspect metadata, roles, and locking configuration.
 * 3. Read records individually and through pagination.
 * 4. Query the record-count and lock-status helpers.
 */

import {
    CapabilityIssueOptions,
    Data,
    LockingConfig,
    LockingWindow,
    PermissionSet,
    TimeLock,
} from "@iota/audit-trail/node";
import { strict as assert } from "assert";
import { getFundedClient, TEST_GAS_BUDGET } from "./util";

export async function accessReadOnlyMethods(): Promise<void> {
    console.log("=== Audit Trail: Read-Only Inspection ===\n");

    // `adminClient` creates the trail and delegates one record write.
    // `recordAdminClient` adds the follow-up record.
    const adminClient = await getFundedClient();
    const recordAdminClient = await getFundedClient();

    const { output: createdTrail } = await adminClient
        .createTrail()
        .withTrailMetadata("Operations Trail", "Used to inspect read-only accessors")
        .withUpdatableMetadata("Status: Active")
        .withLockingConfig(
            new LockingConfig(LockingWindow.withCountBased(BigInt(2)), TimeLock.withNone(), TimeLock.withNone()),
        )
        .withInitialRecordString("Initial record", "event:created")
        .finish()
        .withGasBudget(TEST_GAS_BUDGET)
        .buildAndExecute(adminClient);

    const trailId = createdTrail.id;

    const recordAdminRole = adminClient.trail(trailId).access().forRole("RecordAdmin");
    await recordAdminRole
        .create(PermissionSet.recordAdminPermissions())
        .withGasBudget(TEST_GAS_BUDGET)
        .buildAndExecute(adminClient);
    await recordAdminRole
        .issueCapability(new CapabilityIssueOptions(recordAdminClient.senderAddress()))
        .withGasBudget(TEST_GAS_BUDGET)
        .buildAndExecute(adminClient);

    // RecordAdmin adds a follow-up record.
    await recordAdminClient
        .trail(trailId)
        .records()
        .add(Data.fromString("Follow-up record"), "event:updated")
        .withGasBudget(TEST_GAS_BUDGET)
        .buildAndExecute(recordAdminClient);

    // All reads below require no capability — any address can inspect the trail.
    const onChainTrail = await adminClient.trail(trailId).get();
    console.log("Trail summary:");
    console.log("  id =", onChainTrail.id);
    console.log("  creator =", onChainTrail.creator);
    console.log("  created_at =", onChainTrail.createdAt);
    console.log("  sequence_number =", onChainTrail.sequenceNumber);
    console.log("  immutable_metadata =", onChainTrail.immutableMetadata);
    console.log("  updatable_metadata =", onChainTrail.updatableMetadata, "\n");

    console.log("Roles:", onChainTrail.roles.roles.map((r) => r.name));
    console.log("Locking config:", onChainTrail.lockingConfig, "\n");

    const trailHandle = adminClient.trail(trailId);
    const count = await trailHandle.records().recordCount();
    const initialRecord = await trailHandle.records().get(0n);
    const firstPage = await trailHandle.records().listPage(undefined, 10);
    const recordZeroLocked = await trailHandle.locking().isRecordLocked(0n);

    console.log("Record count:", count);
    console.log("Record #0:", initialRecord);
    console.log("First page size:", firstPage.records.length, "(has_next_page =", firstPage.hasNextPage, ")");
    console.log("Is record #0 locked?", recordZeroLocked);

    assert.equal(count, 2n);
    assert.equal(initialRecord.data.toString(), "Initial record");
    assert.equal(firstPage.records.length, 2);
}

File: examples/audit-trail/07_access_read_only_methods.rs
// Copyright 2020-2026 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

//! ## Actors
//!
//! - **Admin client**: Creates the trail and sets up the RecordAdmin role.
//! - **Record admin client**: Adds one follow-up record. All subsequent operations are read-only and can be performed
//!   by any address — no capability required.

use anyhow::{Result, ensure};
use audit_trail::core::types::{
    CapabilityIssueOptions, Data, ImmutableMetadata, InitialRecord, LockingConfig, LockingWindow, PermissionSet,
    TimeLock,
};
use examples::get_funded_audit_trail_client;
use product_common::core_client::CoreClient;

/// Demonstrates how to:
/// 1. Load the full on-chain trail object.
/// 2. Inspect metadata, roles, and locking configuration.
/// 3. Read records individually and through pagination.
/// 4. Query the record-count and lock-status helpers.
#[tokio::main]
async fn main() -> Result<()> {
    println!("=== Audit Trail: Read-Only Inspection ===\n");

    // Use separate clients to keep write delegation distinct from read-only inspection.
    let admin_client = get_funded_audit_trail_client().await?;
    let record_admin_client = get_funded_audit_trail_client().await?;

    let created_trail = admin_client
        .create_trail()
        .with_trail_metadata(ImmutableMetadata::new(
            "Operations Trail".to_string(),
            Some("Used to inspect read-only accessors".to_string()),
        ))
        .with_updatable_metadata("Status: Active")
        .with_locking_config(LockingConfig {
            delete_record_window: LockingWindow::CountBased { count: 2 },
            delete_trail_lock: TimeLock::None,
            write_lock: TimeLock::None,
        })
        .with_initial_record(InitialRecord::new(
            Data::text("Initial record"),
            Some("event:created".to_string()),
            None,
        ))
        .finish()
        .build_and_execute(&admin_client)
        .await?
        .output;

    let trail_id = created_trail.trail_id;
    let record_admin_role = "RecordAdmin";

    admin_client
        .trail(trail_id)
        .access()
        .for_role(record_admin_role)
        .create(PermissionSet::record_admin_permissions(), None)
        .build_and_execute(&admin_client)
        .await?;
    admin_client
        .trail(trail_id)
        .access()
        .for_role(record_admin_role)
        .issue_capability(CapabilityIssueOptions {
            issued_to: Some(record_admin_client.sender_address()),
            valid_from_ms: None,
            valid_until_ms: None,
        })
        .build_and_execute(&admin_client)
        .await?;

    record_admin_client
        .trail(trail_id)
        .records()
        .add(Data::text("Follow-up record"), Some("event:updated".to_string()), None)
        .build_and_execute(&record_admin_client)
        .await?;

    let on_chain_trail = admin_client.trail(trail_id).get().await?;
    println!(
        "Trail summary:\n  id = {}\n  creator = {}\n  created_at = {}\n  sequence_number = {}\n  immutable_metadata = {:?}\n  updatable_metadata = {:?}\n",
        on_chain_trail.id.object_id(),
        on_chain_trail.creator,
        on_chain_trail.created_at,
        on_chain_trail.sequence_number,
        on_chain_trail.immutable_metadata,
        on_chain_trail.updatable_metadata
    );

    println!(
        "Roles: {:?}\nLocking config: {:?}\n",
        on_chain_trail.roles.roles.keys().collect::<Vec<_>>(),
        on_chain_trail.locking_config
    );

    let read_only_trail = admin_client.trail(trail_id);
    let record_count = read_only_trail.records().record_count().await?;
    let initial_record = read_only_trail.records().get(0).await?;
    let first_page = read_only_trail.records().list_page(None, 10).await?;
    let record_zero_locked = read_only_trail.locking().is_record_locked(0).await?;

    println!("Record count: {record_count}");
    println!("Record #0: {:?}", initial_record);
    println!(
        "First page size: {} (has_next_page = {})",
        first_page.records.len(),
        first_page.has_next_page
    );
    println!("Is record #0 locked? {record_zero_locked}");

    ensure!(record_count == 2);
    ensure!(matches!(initial_record.data, Data::Text(ref text) if text == "Initial record"));
    ensure!(first_page.records.len() == 2);

    Ok(())
}
